### PR TITLE
[WIP] Initial stab at refactor

### PR DESCRIPTION
**Pull request by carnage**
Started to make the token object immutable.

### CODE REVIEW

File: src/JsonToken.php [hunk 1]
```diff
@@ -465,36 +394,29 @@ public function toString(): string
...
         }
+        $key = $this->key;
         /** @var ProtocolInterface $protocol */
         switch ($this->purpose) {
-            case 'auth':
-                if ($this->key instanceof SymmetricAuthenticationKey) {
-                    return $protocol::auth($claims, $this->key, $this->footer);
+            case self::PURPOSE_AUTH:
+                /** @var SymmetricAuthenticationKey $key */
+                return $protocol::auth($claims, $key, $this->footer);
```
Comment (carnage) [inline src/JsonToken.php]: Wondering if this should be wrapped in a try/catch as well

**Comment by paragonie-scott** [comment]
Immutability is a good idea. 👍

**Comment by paragonie-scott** [comment]
I think I'm going to implement where this pull request was going, except with the new local/remote paradigm.

**Comment by carnage** [comment]
I was using this as a reference https://github.com/lcobucci/jwt/blob/master/src/Token/Plain.php for refactoring efforts

**Comment by paragonie-scott** [comment]
The code has diverged pretty significantly since this PR was started. I liked the idea you proposed, but since the merge conflicts keep stacking up I'm going to close it. Thanks @carnage for the immutability suggestion. 👍

**Pull request closed without merging**